File: solutions/0007.Reverse_Integer/solution.go
package solution

const MAX_INT32 = 1<<31 - 1
const MIN_INT32 = 1 << 31

func reverse(x int) int {
	negative := false

	if x < 0 {
		negative = true
		x = -x
	}

	reverse := 0

	for x != 0 {
		mod := x % 10

		if negative {
			if (MIN_INT32-mod)/10 < reverse {
				return 0
			}
		} else {
			if (MAX_INT32-mod)/10 < reverse {
				return 0
			}
		}
		reverse = reverse*10 + mod
		x /= 10
	}

	if negative {
		return -reverse
	}

	return reverse
}

File: solutions/0007.Reverse_Integer/Solution.java
class Solution {
    public int reverse(int x) {
        if (x == Integer.MIN_VALUE) {
            return 0;
        }

        int sign = 1;
        int reverse = 0;

        if (x < 0) {
            sign = -1;
            x = -x;
        }

        while (x != 0) {
            if (reverse * 10.0 + x % 10 > Integer.MAX_VALUE) {
                return 0;
            }

            reverse = reverse * 10 + x % 10;
            x /= 10;
        }

        return reverse * sign;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.reverse(-2147483648));
    }
}
